File: lec01/hw2_large.cc
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <cctype>

using namespace std;

const int SCORES[26] = {1, 3, 2, 2, 1, 3, 3, 1, 1, 4, 4, 2, 2, 1, 1, 3, 4, 1, 1, 1, 2, 3, 3, 4, 3, 4};

// 単語と文字カウントを保持する構造体
struct WordHash {
    string word;
    unordered_map<char, int> char_count;
};

// 文字カウントを計算
unordered_map<char, int> count_characters(const string &word) {
    unordered_map<char, int> char_count;
    for (char c : word) {
        if (isalpha(c)) {
            char_count[tolower(c)]++;
        }
    }
    return char_count;
}

// アナグラムかどうかを判定
bool is_anagram(const unordered_map<char, int> &input_count, const unordered_map<char, int> &word_count) {
    for (const auto &entry : word_count) {
        if (input_count.at(entry.first) < entry.second) {
            return false;
        }
    }
    return true;
}

// スコアを計算
int calculate_score(const unordered_map<char, int> &char_count) {
    int score = 0;
    for (const auto &entry : char_count) {
        score += SCORES[entry.first - 'a'] * entry.second;
    }
    return score;
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        cout << "Usage: ./hw2_large large.txt" << endl;
        return 1;
    }

    string input_file = argv[1];
    string dictionary_file = "words.txt";
    string output_file = "large_answer.txt";

    // 辞書を読み込む
    ifstream dict_fp(dictionary_file);
    if (!dict_fp) {
        cerr << "Failed to open words.txt" << endl;
        return 1;
    }

    vector<WordHash> word_hash_list;
    string line;
    while (getline(dict_fp, line)) {
        WordHash wh;
        wh.word = line;
        wh.char_count = count_characters(line);
        word_hash_list.push_back(wh);
    }
    dict_fp.close();

    // 入力ファイルを読み込む
    ifstream input_fp(input_file);
    if (!input_fp) {
        cerr << "Failed to open input file" << endl;
        return 1;
    }

    ofstream output_fp(output_file);
    if (!output_fp) {
        cerr << "Failed to open output file" << endl;
        input_fp.close();
        return 1;
    }

    while (getline(input_fp, line)) {
        unordered_map<char, int> input_char_count = count_characters(line);

        int max_score = 0;
        string best_word;

        for (const auto &wh : word_hash_list) {
            if (is_anagram(input_char_count, wh.char_count)) {
                int score = calculate_score(wh.char_count);
                if (score > max_score) {
                    max_score = score;
                    best_word = wh.word;
                }
            }
        }

        if (!best_word.empty()) {
            output_fp << best_word << endl;
        }
    }

    input_fp.close();
    output_fp.close();

    return 0;
}

File: lec01/hw2_large.py
# python hw2_large.py large.txt
#  python3 score_checker.py large.txt large_answer.txt

import sys

# SCORES of the characters:
# ----------------------------------------
# | 1 point  | a, e, h, i, n, o, r, s, t |
# | 2 points | c, d, l, m, u             |
# | 3 points | b, f, g, p, v, w, y       |
# | 4 points | j, k, q, x, z             |
# ----------------------------------------


SCORES = [1, 3, 2, 2, 1, 3, 3, 1, 1, 4, 4, 2, 2, 1, 1, 3, 4, 1, 1, 1, 2, 3, 3, 4, 3, 4]


# スコアの計算
def score(word):
    score = 0
    for c in word:
        score += SCORES[ord(c) - ord('a')]
    return score

# 26個の要素を持つ配列(ベクトル)を用意。(26文字でサイズが固定できるので)
def to_vector(word):
    vec = [0]*26
    for c in word:
       vec[ord(c) - ord('a')] += 1
    return tuple(vec)  # リストよりもtuple(組み合わせ)の方がメモリ使用量が少ないらしいので、tupleにして返す


# 引数の確認
if len(sys.argv) < 2:
    print("Usage: python hw2_large.py large.txt")
    sys.exit(1)


with open('words.txt') as dictionary:
    dict_vec = []
    for word in dictionary:
        word = word.strip()
        vec = to_vector(word) 
        dict_vec.append((word,vec))  # dict_vecに単語とベクトルを持たせる

dict_vec.sort(key=lambda x: -score(x[0]))  #スコアが大きい順にソートする


input_file = sys.argv[1] 
with open(input_file,'r') as f:
    anagram_word_list = [] # ここにanagramを保存していく
    for word in f:
        word = word.strip()
        vec = to_vector(word)
        max_score = 0
        for word_dict,dict_word_vec in dict_vec:
            if all(vec[i] >= dict_word_vec[i] for i in range(26)): #辞書の単語の方が、アルファベットの出現回数が少ない時ok
                    max_score = score(word_dict)
                    max_word = word_dict
                    anagram_word_list.append(max_word) # スコア順にソートしているので最初にokだった単語をそのまま追加
                    break
           
  
             
with open('large_answer.txt','w') as o:
    for row in anagram_word_list: # anagram_word_listにある単語を書き込む
        o.write(row + '\n')

        
    
    
